### ISSUE TITLE
case object

### ISSUE THREAD

**Issue by julien-truffaut** [miniboxing/miniboxing-plugin#184]
``` scala
sealed trait Fruit
case object Apple  extends Fruit
case object Orange extends Fruit

val apples: ConsList[Apple.type] = ConsList(Apple, Apple)
```

generate the following error:

```
Error:(111, 17) [ occured while creating miniboxed method apply_J ]
Apple is already defined as (compiler-generated) case class companion object Apple
    case object Apple  extends Fruit
                ^
```

**Comment by VladUreche** [comment]
Thanks for the report @julien-truffaut! I will check out the scalacheck file and see what's happening.
(as far as I can tell, there's some interaction between scalacheck and miniboxing, as the code compiles well in isolation)

**Comment by VladUreche** [comment]
Took crazy-long, but here's a minimization:

``` scala
$ cat gh-bug-184.scala 
package miniboxing.tests.bug184

class ConsListSpec {

  test("widen") {
    sealed trait Fruit
    case object Apple  extends Fruit
    case object Orange extends Fruit

    val apples: List[Apple.type] = List(Apple, Apple)
  }

  def test(name: String)(f: => Unit) = ???
}

gh-bug-184.scala:7: error: [ occured while creating miniboxed method apply_J ]
Apple is already defined as (compiler-generated) case class companion object Apple
    case object Apple  extends Fruit
                ^
one error found
```

**Comment by VladUreche** [comment]
Since it's quite late, I will leave this bug open.
@julien-truffaut: When you add `-P:minibox:warn-off`, it looks like this bug doesn't occur anymore. I forked `brique` (branch `miniboxed`) and was able to run all tests, even this one which you commented.

**Comment by VladUreche** [comment]
Reduced:

``` scala
$ cat gh-bug-184.scala 
package miniboxing.tests.bug184

class ConsListSpec {

  test {
    case object Apple
  }

  def test(f: => Unit) = ???
}

$ mb-scalac gh-bug-184.scala 
gh-bug-184.scala:6: error: [ occured while creating miniboxed method apply_J ]
Apple is already defined as (compiler-generated) case class companion object Apple
    case object Apple
                ^
one error found

```

And another issue, related:

``` scala
$ cat gh-bug-184-2.scala 
package miniboxing.tests.bug184

class ConsListSpec {

  test {
    object Apple
  }

  def test(f: => Unit) = ???
}

$ mb-scalac gh-bug-184-2.scala 
gh-bug-184-2.scala:7: error: [ occured while creating miniboxed method apply_J ]
illegal inheritance;
 self-type <root>.type does not conform to Object's selftype Object
  }
  ^
one error found
```

**Comment by VladUreche** [comment]
Related issue in scalac: https://issues.scala-lang.org/browse/SI-8122

**Comment by VladUreche** [comment]
Okay, I have a fix that works for 2.11. Porting it to 2.10 is too tedious, so I will not even attempt that.

**Comment by VladUreche** [comment]
Done, but this bug alone took about 6 hours overall, so the others will have to wait.
